### PR TITLE
Refactor: builder wrapper

### PR DESCRIPTION
**Pull request by TGlide**
## Description
This PR refactors the internals of Melt UI to use a function called `builder` to create its element-related derived stores and actions, to achieve consistency throughout the project.

I'm opening this as a draft, so it can be reviewed while its being worked upon, as to be able to raise discussions over its implementation and avoid unnecessary rewrites.

As this also addresses #92, we need to be extremely careful as to not introduce bugs with this change. @huntabyte would really appreciate your review, especially in the `Menu` related builders.

### CODE REVIEW

**Comment by huntabyte** [comment]
Made some adjustments with the menu selectors but all seems to be working as expected (functionality-wise).

The actions aren't hidden for the items, which I assume is because we aren't filtering it out when just using the `lightable` for the cases where there aren't any store dependencies for the element like we are with the `hiddenAction` for those that do. I didn't touch that part as I don't want to step on what you already have going, just fixed up the menus :)

Other than that it seems to work well and I do like the builder-builder better than the previous one-offs we had everywhere, incredible work so far dude!

File: src/lib/internal/helpers/builder.ts [hunk 1]
```diff
@@ -103,228 +93,134 @@ export function effect<S extends Stores>(
...
+						};
+					};
+					fn.action = action ?? noop;
+					return fn;
+				}
+
+				return hiddenAction({
+					...result,
+					[`data-melt-${name}`]: '',
+					action: action ?? noop,
```
Comment (huntabyte) [inline src/lib/internal/helpers/builder.ts:164]: Here we're returning a hidden action if there are `stores`, which is great, as expected

File: src/lib/internal/helpers/builder.ts [hunk 2]
```diff
@@ -103,228 +93,134 @@ export function effect<S extends Stores>(
...
+			if (isFunctionWithParams(result)) {
+				const resultFn = (...args: Parameters<typeof result>) => {
+					return {
+						...result(...args),
+						[`data-melt-${name}`]: '',
+					};
+				};
+				resultFn.action = action ?? noop;
+
+				return lightable(resultFn);
```
Comment (huntabyte) [inline src/lib/internal/helpers/builder.ts:179]: But here we're just returning a `lightable`, which doesn't appear to filter out the action so it still shows the entire action string.

File: src/lib/internal/helpers/builder.ts [hunk 3]
```diff
@@ -103,228 +93,134 @@ export function effect<S extends Stores>(
...
 
-		return { attach, getElement, addAction, index: index++ };
-	};
-
-	const addUnsubscriber: AddUnsubscriber = (cb) => {
-		if (Array.isArray(cb)) {
-			unsubscribers.push(...(cb.filter((a) => a !== undefined) as (() => void)[]));
-		} else {
-			unsubscribers.push(cb);
+			return lightable({
```
Comment (huntabyte) [inline src/lib/internal/helpers/builder.ts:182]: And here as well. So possibly checking if there is an action and applying the same logic as `hiddenAction` to the lightable would be ideal?

**Review by huntabyte — COMMENTED** [review]
Great work man! Loving the new syntax... much cleaner than `.action`! Just the issue with the action strings not being filtered but other than that everything appears to work as expected and I believe it's more intuitive than the previous way we were handling things.

**Comment by TGlide** [comment]
> Made some adjustments with the menu selectors but all seems to be working as expected (functionality-wise).
> 
> The actions aren't hidden for the items, which I assume is because we aren't filtering it out when just using the `lightable` for the cases where there aren't any store dependencies for the element like we are with the `hiddenAction` for those that do. I didn't touch that part as I don't want to step on what you already have going, just fixed up the menus :)
> 
> Other than that it seems to work well and I do like the builder-builder better than the previous one-offs we had everywhere, incredible work so far dude!

Oops, that's my bad!

File: src/lib/internal/helpers/builder.ts (hunk 3 above)
Comment (TGlide) [reply src/lib/internal/helpers/builder.ts:182]: Yuo!

File: src/lib/internal/helpers/builder.ts (hunk 2 above)
Comment (TGlide) [reply src/lib/internal/helpers/builder.ts:179]: It's not needed here, as the fn doesn't return an action. Same as on the derived above.

**Pull request merged**